src/prefect/infrastructure/provisioners/container_instance.py

```diff
@@ -0,0 +1,765 @@
+        )
+        await self.azure_cli.run_command(
+            resource_group_command,
+            success_message=(
+                f"Resource group '{self.RESOURCE_GROUP_NAME}' created successfully"
+            ),
+            failure_message=(
+                f"Failed to create resource group '{self.RESOURCE_GROUP_NAME}' in"
+                f" subscription '{self._subscription_name}'"
+            ),
+            ignore_if_exists=True,
+        )
+
+    async def _create_app_registration(self) -> str:
+        """
+        Creates an app registration in Azure Active Directory.
+
+        Returns:
+            str: The client ID of the newly created app registration.
+
+        Raises:
+            subprocess.CalledProcessError: If the Azure CLI command execution fails.
+        """
+        # Check if the app registration already exists
+        check_exists_command = (
+            f"az ad app list --display-name {self.APP_REGISTRATION_NAME} --output json"
+        )
+        _, app_registrations = await self.azure_cli.run_command(
+            check_exists_command,
+        )
+        if app_registrations:
+            if isinstance(app_registrations, str):
+                app_registrations = json.loads(app_registrations)
+
+            existing_app_registration = next(
+                (
+                    app
+                    for app in app_registrations
+                    if app["displayName"] == self.APP_REGISTRATION_NAME
+                ),
+                None,
+            )
+            if existing_app_registration:
+                self._console.print(
+                    f"App registration '{self.APP_REGISTRATION_NAME}' already exists.",
+                    style="yellow",
+                )
+                return existing_app_registration["appId"]
+
+        app_registration_command = (
+            f"az ad app create --display-name {self.APP_REGISTRATION_NAME} "
+            "--output json"
+        )
+        result, app_registration = await self.azure_cli.run_command(
+            app_registration_command,
+            success_message=(
+                f"App registration '{self.APP_REGISTRATION_NAME}' created successfully"
+            ),
+            failure_message=(
+                "Failed to create app registration with name"
+                f" '{self.APP_REGISTRATION_NAME}'"
+            ),
+            ignore_if_exists=True,
+        )
+        if result == "created":
+            if app_registration:
+                if isinstance(app_registration, str):
+                    app_registration = json.loads(app_registration)
+                return app_registration["appId"]
+
+        elif result == "exists":
+            self._console.print(
+                f"App registration '{self.APP_REGISTRATION_NAME}' already exists.",
+                style="yellow",
+            )
+        elif result == "error":
+            raise Exception("Error creating the app registration.")
```

Comment:
Looks like there are two different modes of operation here:

1. You run a separate command to check if something exists
2. You let `AzureCLI` let you know if something already existed

This doesn't need to be addressed in this PR, but having a consistent sequence for each thing getting created would be nice. My guess is the different approaches are probably due to the idiosyncrasies of the `az` CLI, but it'd be nice to find a way to hide that implementation detail.

Reply:
That makes sense. I had started in one direction (mode 2) but ended up converting to mode 1 as it felt more thorough. This should be consistent across the board now in [0694586](https://github.com/PrefectHQ/prefect/pull/11275/commits/0694586696b0a3c81c23fa6b477234b743a8d730)